[Reviewer: Rob] Add GR support to Chronos

eleanor-merry opened the pull request:
Rob, can you please review this PR to add GR support to Chronos. This PR includes the code, UTs, FVs and docs.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metaswitch/chronos/284)
<!-- Reviewable:end -->

rkd-msw commented:
Looks good. I think the three areas I really care about are:

* efficiency - there are a couple of nested for loops, and I worry that this will be inefficient when we're replicating to more places
* test readability - I think Google Test's container-based matchers would be helpful
* behaviour when we have a replication factor of 3 but only two Chronos nodes in a particular site - I think this is might be a tricky-to-diagnose bug

---

Reviewed 31 of 31 files at r1.
Review status: all files reviewed at latest revision, 41 unresolved discussions.

---

*[doc/api.md, line 165 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPln9fTtEJmm4F2PvOJ:-KPln9fTtEJmm4F2PvOK:bmmvlwc) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/doc/api.md#L165)):*
> ```Markdown
>  * `reliability/cluster-view-id` - The ID of the current cluster topology (used for error-checking and to spot inconsistencies in the cluster)
>  * `reliability/replicas` - The ordered list of the replicas for the timer, the receiving node can use this to work out when it should pop the timer.
>  * `reliability/sites` - The ordered list of the sites for the timer, the receiving node can use this to work out when it should pop the timer.
> ```

Is it worth noting where the names come from (e.g. "matching the names in /etc/chronos/chronos.conf")? Maybe the same for 'replicas'.

---

*[doc/configuration.md, line 56 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPlzzYazCjQsYt7Rxi5:-KPlzzYbL5_qxszreo0C:b-trlcgc) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/doc/configuration.md#L56)):*
> ```Markdown
>     [sites]
>     local_site = local-site-name             # The name of the local site
>     remote_site = site-b=bar.foo.com:8000    # The name of a remote site, and the address
> ```

Should we be more precise than 'address of the remote site', e.g. 'a DNS name resolving to all Chronos nodes in the remote site'?

---

*[doc/gr.md, line 5 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPlq2FDRFMH_p0fiMfC:-KPlq2FEIFRey1AL2WS5:b-n64ssa) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/doc/gr.md#L5)):*
> ```Markdown
> Geographic redundancy (GR) is supported in Chronos by using essentially the same methods as we ensure in-site redundancy. These are described [here](https://github.com/Metaswitch/chronos/blob/dev/doc/technical.md).
> 
> At a high level, when a timer is added (or first modified after a deployment becomes GR) it creates a site ordering list. This list simply takes the available sites, makes the local site the primary site, and then orders the rest of the sites randomly after it. Chronos knows what sites are available as these are configured in the [GR config](https://github.com/Metaswitch/chronos/blob/dev/doc/configuration.md).
> ```

Presumably the site ordering list also changes when a timer is modified after adding a site? (e.g. not just becoming GR, but moving from 2-site to 3-site)

---

*[doc/gr.md, line 7 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPlpoABTF_23kvxSyi4:-KPlpoABTF_23kvxSyi5:b5hls1w) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/doc/gr.md#L7)):*
> ```Markdown
> At a high level, when a timer is added (or first modified after a deployment becomes GR) it creates a site ordering list. This list simply takes the available sites, makes the local site the primary site, and then orders the rest of the sites randomly after it. Chronos knows what sites are available as these are configured in the [GR config](https://github.com/Metaswitch/chronos/blob/dev/doc/configuration.md).
> 
> The Chronos process then replicates the timer both within site and cross site. Each Chronos process that the timer is replicated to uses both the site's position in the site ordering list and the Chronos node's position in the replica list to decide what offset to apply to the pop time of the timer. For example, in a 3 site deployment with two replicas per site you can get a timer with the following replicas and pop delays.
> ```

Is there a simpler way to say "uses both the site's position in the site ordering list and the Chronos node's position in the replica list to decide what offset to apply to the pop time of the timer"? "delays the timer based on both the site's priority and its own priority within the site", maybe?

---

*[doc/gr.md, line 12 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPlo13zD7leXXgh2sMS:-KPlo13zD7leXXgh2sMT:b-ebokjb) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/doc/gr.md#L12)):*
> ```Markdown
>  * Replica 1 - No delay
>  * Replica 2 - 2 sec delay
> * Site 3
> ```

Is it worth being explicit that site 3 has been chosen as first backup and site 2 as second backup? (I initially wondered if this was a typo.)

---

*[doc/gr.md, line 21 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPluF1mlZ9tcLTuRNod:-KPluF1mlZ9tcLTuRNoe:bv3nifs) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/doc/gr.md#L21)):*
> ```Markdown
> When a timer pops in a Chronos process, Chronos handles the timer as normal (e.g. handling the callback, replicating the timer/tombstone to all other replicas of the timer).
>  
> This solution provides the desired data resiliency. A timers can be delayed by `((number of sites - 1) * number of replicas * 2) + 2 + site latencies)` seconds in the case of a multi-site failure.
> ```

"A timers"

---

*[doc/gr.md, line 39 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPlv19xVJKIbuOkMdJu:-KPlv19xVJKIbuOkMdJv:b-4n2lu4) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/doc/gr.md#L39)):*
> ```Markdown
> All new timers include the failed sites when they calculate the site priority list for the timer, but sole working site should always be the primary site (as the primary site is always the site that creates the timer). Therefore, new timers won't have delayed pop times.
> 
> ### Net split between site
> ```

"between sites"?

---

*[doc/gr.md, line 43 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPlv2mudkkOCwDUC1Hu:-KPlv2mudkkOCwDUC1Hv:b-sagmu) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/doc/gr.md#L43)):*
> ```Markdown
> When the communication between sites goes down, rather than a site going down, then timer replication requests won't be sent successfully between sites.
> 
> Existing timers will double pop (e.g. if `site1` is the primary site it will pop at 0 secs, then `site2` will pop at 4 secs). All users of Chronos should be safe to this type of double pop.
> ```

Worth noting that this double-pop-resilience requirement already exists, to deal with netsplits within a site?

---

*[include/chronos_gr_connection.h, line 44 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPlw-3mQKJq9GWrEJuc:-KPlw-3mQKJq9GWrEJud:b-r91fws) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/include/chronos_gr_connection.h#L44)):*
> ```objc
> /// @class ChronosGRConnection
> ///
> /// Responsible for sending replication requests between sites
> ```

It looks like each ChronosGRConnection is responsible for GR replication to one other site - worth saying so?

---

*[include/globals.h, line 110 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPm-L71aKrA9xY2bxvx:-KPm-L71aKrA9xY2bxvy:bf1p2a1) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/include/globals.h#L110)):*
> ```objc
>   GLOBAL(remote_sites, std::map<std::string, std::string>);
>   GLOBAL(remote_site_names, std::vector<std::string>);
>   GLOBAL(remote_site_hosts, std::vector<std::string>);
> ```

'_hosts' feels a bit misleading (because it's not all the hosts in the remote site, it's a DNS record through which they can be reached).

I'm also slightly surprised that we have different vectors, even though there should presumably be exactly one hostname per site name - better as a map or a vector of pairs?

---

*[include/timer.h, line 155 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmQdIVd_xlpm6dAnnJ:-KPmQdIVd_xlpm6dAnnK:bteldft) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/include/timer.h#L155)):*
> ```objc
>   uint32_t delay_from_site_position() const;
> 
>   // Work out how delayed the timer should be based on the timer's sequence
> ```

Worth having an example ("e.g. if this timer is set to pop every 20 seconds for 100 seconds") - I forgot what a sequence number was, I think because I was in the mindset of cross-site replication and assumed it was to do with replica ordering.

---

*[scripts/chronos_fv_test.py, line 122 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPm7sikxIokfgKnxyAF:-KPm7sikxIokfgKnxyAG:bjpxchf) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/scripts/chronos_fv_test.py#L122)):*
> ```Python
>     for i in range(lower, upper):
>         Popen([CHRONOS_BINARY, '--daemon', '--pidfile', PID_PATTERN % i, '--config-file',
>             CONFIG_FILE_PATTERN % i, '--cluster-config-file',
> ```

Can you break the lines differently so `'--cluster-config-file', CLUSTER_CONFIG_FILE_PATTERN % i` are on the same line?

---

*[scripts/chronos_fv_test.py, line 231 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPm8NOhuPOy79YB5ntv:-KPm8NOij_bi4S2tw4qM:b-bzvsxr) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/scripts/chronos_fv_test.py#L231)):*
> ```Python
>         [sites]
>         local_site = {local_site}
>         """).format(**locals()))
> ```

Do you need `**locals()`? `"""local_site = {}""".format(local_site)` feels simpler.

---

*[scripts/chronos_fv_test.py, line 291 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPm8jkdVMTKF4mGDrOe:-KPm8jkdVMTKF4mGDrOf:b-sh7ja6) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/scripts/chronos_fv_test.py#L291)):*
> ```Python
>             write_conf(CONFIG_FILE_PATTERN % num,
>                        chronos_nodes[num])
>             write_cluster_conf(CLUSTER_CONFIG_FILE_PATTERN % num,
> ```

Maybe worth using keyword arguments for readability? I can't immediately tell what the empty lists are.

---

*[scripts/chronos_gr.py, line 58 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmSaYa4OeW1naSPTIP:-KPmSaYb9Hi5n6z5k4d9:bj3cxg9) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/scripts/chronos_gr.py#L58)):*
> ```Python
> 
>         # Check that all the timers have popped
>         sleep(12)
> ```

Why are the sleeps the durations they are? Worth a comment.

---

*[scripts/chronos_gr.py, line 73 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmSrD5ukiPHjcI2Kn2:-KPmSrD5ukiPHjcI2Kn3:bwu3rod) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/scripts/chronos_gr.py#L73)):*
> ```Python
>         self.write_gr_config_for_nodes([3], 'site4', ['site1=127.0.0.11:7253','site2=127.0.0.12:7254','site3=127.0.0.13:7255'])
>         start_nodes(0, 4)
>         create_timers(chronos_nodes[0], 100)
> ```

Might be worth creating 25 timers in each site, just to check that that works - I could imagine some bug where replication from the first site worked but replication from others didn't.

---

*[scripts/chronos_gr.py, line 78 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmTCPNi0EGv3ZzyOBz:-KPmTCPNi0EGv3ZzyOC-:b-5sor25) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/scripts/chronos_gr.py#L78)):*
> ```Python
>         # Now kill all but the last site, and check that all timers pop within
>         # expected delay
>         kill_nodes(0, 3)
> ```

I'd be mildly tempted to kill three randomly-chosen sites (just to catch bugs where replication from B to D works but B to C doesn't).

---

*[src/chronos_gr_connection.cpp, line 47 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmP_toAa1kLovwr6RR:-KPmP_toAa1kLovwr6RS:brt2xv1) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/chronos_gr_connection.cpp#L47)):*
> ```cpp
>                            false,
>                            resolver,
>                            SASEvent::HttpLogLevel::DETAIL,
> ```

Is this right? I think we use ::NONE for the within-site replication.

---

*[src/gr_replicator.cpp, line 87 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmPreFUbGMBb3IWbkL:-KPmPreFUbGMBb3IWbkM:bwi79s1) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/gr_replicator.cpp#L87)):*
> ```cpp
> {
>   // Create the JSON body - strip out any replica information
>   Timer* timer_copy = new Timer(*timer);
> ```

Should you just create this timer on the stack, if you're deleting it at the end of the function?

---

*[src/gr_replicator.cpp, line 90 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmQ0rrOwgUM0g6ZFRL:-KPmQ0rrOwgUM0g6ZFRM:becv62y) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/gr_replicator.cpp#L90)):*
> ```cpp
>   std::string url = timer_copy->url();
>   timer_copy->replicas.clear();
>   std::string body = timer_copy->to_json();
> ```

I wonder if it's better to have a with_replicas=false argument to to_json, so you don't need to copy the timer and clear the replicas. Maybe not worth it.

---

*[src/gr_replicator.cpp, line 94 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmQKvGcYvmfHlHaDrU:-KPmQKvHMyXEWjPx7ITG:be8yng6) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/gr_replicator.cpp#L94)):*
> ```cpp
>   for (ChronosGRConnection* conn : _connections)
>   {
>     GRReplicationRequest* replication_request = new GRReplicationRequest();
> ```

Should GRReplicationRequest have a constructor instead of setting values in the struct?

---

*[src/main.cpp, line 325 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPm18zLidtHP7oMOxix:-KPm18zLidtHP7oMOxiy:b-fuhozp) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/main.cpp#L325)):*
> ```cpp
>   // Create the GR connections
>   std::vector<ChronosGRConnection*> gr_connections;
>    std::vector<std::string> remote_site_hosts;
> ```

Indentation?

---

*[src/main.cpp, line 328 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPm1GJw4ruFfmV0qa_C:-KPm1GJw4ruFfmV0qa_D:btqgny9) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/main.cpp#L328)):*
> ```cpp
>   __globals->get_remote_site_hosts(remote_site_hosts);
> 
>   for (std::string site: remote_site_hosts)
> ```

Any reason for this not to be created in the GRReplicator? (I don't mind that much, though.)

---

*[src/timer.cpp, line 129 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmG-oOLFwhyKNMot4K:-KPmG-oOLFwhyKNMot4L:b-2v4hn) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/timer.cpp#L129)):*
> ```cpp
>   __globals->get_local_site_name(local_site_name);
> 
>   for (std::vector<std::string>::const_iterator it = sites.begin();
> ```

Think you could do `site_index = std::find(sites.begin(), sites.end(), local_site_name) - sites.begin()` (see http://stackoverflow.com/a/15099748)

---

*[src/timer.cpp, line 139 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmS2UFhM26NWaclIG5:-KPmS2UFhM26NWaclIG6:b-kny0eb) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/timer.cpp#L139)):*
> ```cpp
>   }
> 
>   return site_index * replicas.size() * DELAY_BETWEEN_CHRONOS_INSTANCES_MS;
> ```

Hmm. What if one site has fewer replicas? e.g. if I want a replication factor of 3, and site A has 4 Chronos nodes, but site B only has 2 for whatever reason.

---

*[src/timer.cpp, line 326 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmGUwHMsVy9ZowGsd1:-KPmGUwIhf6aanCXVzKQ:b7cg5du) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/timer.cpp#L326)):*
> ```cpp
>         writer->StartArray();
>         {
>           for (std::vector<std::string>::iterator it = sites.begin();
> ```

Range-based for loop?

---

*[src/timer.cpp, line 685 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmGuSYrPGM9ce5T73g:-KPmGuSYrPGM9ce5T73h:bn42s4x) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/timer.cpp#L685)):*
> ```cpp
>   // Shuffle the remote sites
>   std::random_shuffle(remote_site_names.begin(), remote_site_names.end());
>   for (std::string site: remote_site_names)
> ```

This feels inefficient - I think it's O(n^2).

---

*[src/timer_handler.cpp, line 196 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPm7ORUMhK6u_JIMCXa:-KPm7ORUMhK6u_JIMCXb:bt89zfe) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/timer_handler.cpp#L196)):*
> ```cpp
> 
>       // Update the site information
>       save_site_information(new_tp.active_timer, existing_tp.active_timer);
> ```

I'm confused, although I think this is probably right and I don't understand the existing terminology here. The comment above says we're adding a new timer, not replacing the existing timer, so why are we saving site information from an existing timer?

---

*[src/timer_handler.cpp, line 715 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPm6tK_ZsUylaUqKZ2R:-KPm6tK_ZsUylaUqKZ2S:b-dtptx4) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/timer_handler.cpp#L715)):*
> ```cpp
>   for (std::string site: old_timer->sites)
>   {
>     if (std::find(new_timer->sites.begin(), new_timer->sites.end(), site) !=
> ```

Should we be using some other data structure here (perhaps a set)? This looks like it's O(n^2) in the number of sites, which is worrying if n gets up to around 10.

---

*[src/ut/test_globals.cpp, line 211 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPm1vR0PzUa0Zgo-Gnk:-KPm1vR0PzUa0Zgo-Gnl:b-d3bqav) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_globals.cpp#L211)):*
> ```cpp
>   EXPECT_EQ(remote_sites["b"], "bar.com");
>   EXPECT_EQ(remote_site_names.size(), 2);
>   EXPECT_TRUE(std::find(remote_site_names.begin(), remote_site_names.end(), "a") != remote_site_names.end());
> ```

Using https://github.com/google/googletest/blob/master/googlemock/docs/CheatSheet.md#container-matchers would be clearer - I might go for `EXPECT_THAT(remote_site_names, ContainerEq({"a", "b"}))`

---

*[src/ut/test_gr_replicator.cpp, line 78 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPm9zB90tC-bQgjS2T3:-KPm9zB90tC-bQgjS2T4:b-l2nih2) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_gr_replicator.cpp#L78)):*
> ```cpp
>   // Timer should have an ID of 1, and a replication factor of 1. If it doesn't
>   // the send_put will fail in the Chronos GR connection
>   fakecurl_responses["http://10.42.42.42:80/timers/0000000000000001-1"] = CURLE_OK;
> ```

Just to check - it's right that we're expecting port 80 here, rather than the bind_port? (I assume that the way defaulting to the bind_port works is that we'd use that port when creating a HttpResolver, but wanted to be sure.)

---

*[src/ut/test_gr_replicator.cpp, line 83 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPm9UnsoB6vjDkLeoE_:-KPm9UnsoB6vjDkLeoEa:b8a1p4x) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_gr_replicator.cpp#L83)):*
> ```cpp
>   _gr->replicate(timer1);
> 
>   // Have to sleep to make sure there's time for the request to have been
> ```

Is it worth adding a UT-only method to wait until the replicator's queue is empty, to avoid needing to sleep (and possibly spurious failures under valgrind on a busy system)?

---

*[src/ut/test_gr_replicator.cpp, line 95 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPm9JDxQSnZXeGv0FSm:-KPm9JDxQSnZXeGv0FSn:bieof3y) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_gr_replicator.cpp#L95)):*
> ```cpp
>   ASSERT_FALSE(doc["reliability"].HasMember("replicas"));
> 
>   std::string error;
> ```

Should you be asserting anything about these variables?

---

*[src/ut/test_handler.cpp, line 466 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmF1d2fW6FWeraYh_Y:-KPmF1d2fW6FWeraYh_Z:b-sug071) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_handler.cpp#L466)):*
> ```cpp
>   EXPECT_EQ(added_timer->sites.size(), 4);
>   EXPECT_EQ(added_timer->sites[0], "local_site_name");
>   EXPECT_TRUE(std::find(added_timer->sites.begin(), added_timer->sites.end(), "remote_site_1_name") != added_timer->sites.end());
> ```

Again, I'd use Google Test's container tests here.

---

*[src/ut/test_handler.cpp, line 472 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmF83jjOZPsC3DvKeP:-KPmF83jjOZPsC3DvKeQ:b524bc1) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_handler.cpp#L472)):*
> ```cpp
> 
>   __globals->lock();
>   __globals->set_remote_site_names(old_remote_site_names);
> ```

As with the comment on another test, I think doing this here will mean other tests fail if this test fails.

---

*[src/ut/test_main.cpp, line 45 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPlz5IvLvFxORZ9dh3U:-KPlz5IvLvFxORZ9dh3V:bommfjb) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_main.cpp#L45)):*
> ```cpp
>   ::testing::InitGoogleMock(&argc, argv);
>   curl_global_cleanup();
>   srand(time(NULL));
> ```

Worth printing out the random seed (and maybe making it overrideable?) so you can reproduce a failing test.

---

*[src/ut/test_timer.cpp, line 319 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmCWlmIm4nezgoz2Vs:-KPmCWlmIm4nezgoz2Vt:b-4f41rr) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_timer.cpp#L319)):*
> ```cpp
> TEST_F(TestTimer, FromJSONTestsBadStatisticsObject)
> {
>   std::string err; bool replicated; bool gr_replicated; Timer* timer;
> ```

Not wild about the semicolon-separated variables on one line.

---

*[src/ut/test_timer.cpp, line 602 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmD_2UgR_VtOkGLDta:-KPmD_2VTegoBEPkYHy3:br9k7pq) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_timer.cpp#L602)):*
> ```cpp
>   t->replicas = replicas;
>   t->start_time_mono_ms = 1000000;
>   t->sequence_number = 0;
> ```

You're setting sequence number twice.

---

*[src/ut/test_timer.cpp, line 628 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmRD38VZUTdil6IFlU:-KPmRD39E5NMUYOK7W6b:bui3y7j) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_timer.cpp#L628)):*
> ```cpp
>   t->sequence_number = 0;
> 
>   EXPECT_EQ(t->next_pop_time(), 1008100);
> ```

Is it worth a comment saying how the 8100 is generated? (I think it's that local_site_name is the second site, there are three replicas per site, and this is the second replica in the second site, so the 5th replica overall, so 0, 2, 4, 6, **8**, and it's a timer to pop 100ms in the future).

---

*[src/ut/test_timer_handler.cpp, line 614 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmAfZiD3fs9pfR2ooD:-KPmAfZiD3fs9pfR2ooE:bl5dujz) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_timer_handler.cpp#L614)):*
> ```cpp
>   // Delete the timer (this is normally done by the insert call, but this
>   // is mocked out)
>   delete timer2;
> ```

Why don't you need to delete timer as well as timer2?

---

*[src/ut/test_timer_handler.cpp, line 1119 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmC4-b3l4iyf2Dc-4p:-KPmC4-b3l4iyf2Dc-4q:bonkavz) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_timer_handler.cpp#L1119)):*
> ```cpp
> 
>   __globals->lock();
>   __globals->set_remote_site_names(old_remote_site_names);
> ```

If this test fails partway through, will it leave the wrong globals set and cause other tests to fail? Should this be done in a teardown block?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/metaswitch/chronos/284#-:-KPmTyr09dsY6aVzIFTK:b-6la68y)*
<!-- Sent from Reviewable.io -->

eleanor-merry commented:
Review status: 8 of 33 files reviewed at latest revision, 41 unresolved discussions.

---

*[doc/gr.md, line 5 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPlq2FDRFMH_p0fiMfC:-KPnXKSlt1v6BR0aF-z7:bjbnf30) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/doc/gr.md#L5)):*
<details><summary><i>Previously, rkd-msw (Rob Day) wrote…</i></summary>
> Presumably the site ordering list also changes when a timer is modified after adding a site? (e.g. not just becoming GR, but moving from 2-site to 3-site)
</details>
Yes - I've not covered that in the high level bit, but it's discussed later in the doc

---

*[include/globals.h, line 110 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPm-L71aKrA9xY2bxvx:-KPnFuKlcsGaQMA5bAJb:b-6e9l2i) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/include/globals.h#L110)):*
<details><summary><i>Previously, rkd-msw (Rob Day) wrote…</i></summary>
> '_hosts' feels a bit misleading (because it's not all the hosts in the remote site, it's a DNS record through which they can be reached).
> 
> I'm also slightly surprised that we have different vectors, even though there should presumably be exactly one hostname per site name - better as a map or a vector of pairs?
</details>
I need both bits of information (name + hostname), but never as a pair. I split them up as two vectors at the start to make the later code simpler. I've changed the name to remote_site_dns_records

---

*[src/chronos_gr_connection.cpp, line 47 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmP_toAa1kLovwr6RR:-KPnVYRuBkS5EOF5zz5C:b-l97i90) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/chronos_gr_connection.cpp#L47)):*
<details><summary><i>Previously, rkd-msw (Rob Day) wrote…</i></summary>
> Is this right? I think we use ::NONE for the within-site replication.
</details>
Good point, fixed

---

*[src/gr_replicator.cpp, line 90 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmQ0rrOwgUM0g6ZFRL:-KPrCUfW81eF4KgLkcuu:b4t6goa) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/gr_replicator.cpp#L90)):*
<details><summary><i>Previously, rkd-msw (Rob Day) wrote…</i></summary>
> I wonder if it's better to have a with_replicas=false argument to to_json, so you don't need to copy the timer and clear the replicas. Maybe not worth it.
</details>
Eh, not convinced - I've not added this option, let me know if that's now an issue

---

*[src/timer.cpp, line 139 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmS2UFhM26NWaclIG5:-KPnGmGSJNJ1_F3bvLrC:b2136ya) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/timer.cpp#L139)):*
<details><summary><i>Previously, rkd-msw (Rob Day) wrote…</i></summary>
> Hmm. What if one site has fewer replicas? e.g. if I want a replication factor of 3, and site A has 4 Chronos nodes, but site B only has 2 for whatever reason.
</details>
Good spot! I've changed this to use the replication factor instead of the replica size (and tweaked how the replication factor gets parsed).

---

*[src/timer_handler.cpp, line 196 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPm7ORUMhK6u_JIMCXa:-KPnIjXVyP_P42jahex6:b-90qe2i) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/timer_handler.cpp#L196)):*
<details><summary><i>Previously, rkd-msw (Rob Day) wrote…</i></summary>
> I'm confused, although I think this is probably right and I don't understand the existing terminology here. The comment above says we're adding a new timer, not replacing the existing timer, so why are we saving site information from an existing timer?
</details>
When we're adding a timer we always pull out any existing timer from the store. We fall into this branch when we're in the case that there was an existing timer in the store with the same ID, and we've decided that the new timer is more up to date, so we're going to add the new timer. We may still want to carry over some information from the existing timer though.

The alternative is deciding that the 'new' timer is actually older than the timer we're currently holding, so we instead want to add the existing timer back to the store. In this case we don't want any information from the 'new' timer

---

*[src/timer_handler.cpp, line 715 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPm6tK_ZsUylaUqKZ2R:-KPrcq5ZZtPMVHq3Vv0C:b5byl9f) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/timer_handler.cpp#L715)):*
<details><summary><i>Previously, rkd-msw (Rob Day) wrote…</i></summary>
> Should we be using some other data structure here (perhaps a set)? This looks like it's O(n^2) in the number of sites, which is worrying if n gets up to around 10.
</details>
We've discussed, and there's no particularly nice way to fix this. I've shuffled round the checks to reduce the time in the mainline cases

---

*[src/ut/test_globals.cpp, line 211 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPm1vR0PzUa0Zgo-Gnk:-KPnPVTWi_OcL7dpR5K6:b-28wwm6) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_globals.cpp#L211)):*
<details><summary><i>Previously, rkd-msw (Rob Day) wrote…</i></summary>
> Using https://github.com/google/googletest/blob/master/googlemock/docs/CheatSheet.md#container-matchers would be clearer - I might go for `EXPECT_THAT(remote_site_names, ContainerEq({"a", "b"}))`
</details>
Ok (I've actually gone for UnorderedElementsAreArray as ContainerEq is order dependant)

---

*[src/ut/test_gr_replicator.cpp, line 78 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPm9zB90tC-bQgjS2T3:-KPnNrFjtTxaXXu3X0nN:bb7866z) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_gr_replicator.cpp#L78)):*
<details><summary><i>Previously, rkd-msw (Rob Day) wrote…</i></summary>
> Just to check - it's right that we're expecting port 80 here, rather than the bind_port? (I assume that the way defaulting to the bind_port works is that we'd use that port when creating a HttpResolver, but wanted to be sure.)
</details>
Yup

---

*[src/ut/test_timer_handler.cpp, line 614 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmAfZiD3fs9pfR2ooD:-KPnRmYJq8lLKE5730n-:bkpuovf) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_timer_handler.cpp#L614)):*
<details><summary><i>Previously, rkd-msw (Rob Day) wrote…</i></summary>
> Why don't you need to delete timer as well as timer2?
</details>
This is deleted by the handler code (it's returned by the fetch call in the insert pair, the timer handler compares it to timer2, decides timer2 is newer so adds timer2 to the store and deletes timer).

---

*[src/ut/test_timer_handler.cpp, line 1119 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmC4-b3l4iyf2Dc-4p:-KPnR1ildXS1MpTDqtMd:b-886gsx) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_timer_handler.cpp#L1119)):*
<details><summary><i>Previously, rkd-msw (Rob Day) wrote…</i></summary>
> If this test fails partway through, will it leave the wrong globals set and cause other tests to fail? Should this be done in a teardown block?
</details>
Fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/metaswitch/chronos/284)*
<!-- Sent from Reviewable.io -->

rkd-msw commented:
Review status: 8 of 33 files reviewed at latest revision, 40 unresolved discussions.

---

*[include/chronos_gr_connection.h, line 45 \[r2\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPrwRT_Qz-FdG140Yt7:-KPrwRT_Qz-FdG140Yt8:b-qpnfvy) ([raw file](https://github.com/metaswitch/chronos/blob/0898686bb88400084e7679ca3ea150a63b454012/include/chronos_gr_connection.h#L45)):*
> ```objc
> ///
> /// Responsible for sending replication requests between sites. Each connection
> /// is responsible for replication to a single remote site.
> ```

foobar

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/metaswitch/chronos/284)*
<!-- Sent from Reviewable.io -->

rkd-msw commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 8 of 33 files reviewed at latest revision, 3 unresolved discussions.

---

*[doc/gr.md, line 7 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPlpoABTF_23kvxSyi4:-KPryLqsjyVhlVVNMY9B:b-4s3g9v) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/doc/gr.md#L7)):*
<details><summary><i>Previously, rkd-msw (Rob Day) wrote…</i></summary>
> Is there a simpler way to say "uses both the site's position in the site ordering list and the Chronos node's position in the replica list to decide what offset to apply to the pop time of the timer"? "delays the timer based on both the site's priority and its own priority within the site", maybe?
</details>
bases on -> based on

---

*[src/gr_replicator.cpp, line 90 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmQ0rrOwgUM0g6ZFRL:-KPrylukM4u0Gpd6FV3h:b-80ayxb) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/gr_replicator.cpp#L90)):*
<details><summary><i>Previously, eleanor-merry (Eleanor Merry) wrote…</i></summary>
> Eh, not convinced - I've not added this option, let me know if that's now an issue
</details>
I'm not that fussed now you've created the timer on the stack.

---

*[src/ut/test_timer_handler.cpp, line 1119 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmC4-b3l4iyf2Dc-4p:-KPs0RW8aARysmpfL_mk:b-xlcvuu) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_timer_handler.cpp#L1119)):*
<details><summary><i>Previously, eleanor-merry (Eleanor Merry) wrote…</i></summary>
> Fixed
</details>
I couldn't immediately see where you'd fixed it, but it looks like the globals were set up in Base::SetUp anyway, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/metaswitch/chronos/284#-:-KPs0fklCVMlf9Kv_IIR:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rkd-msw commented:
Review status: 8 of 33 files reviewed at latest revision, 2 unresolved discussions.

---

*[include/gr_replicator.h, line 54 \[r2\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPs0ny4YdbuvPTrMR5p:-KPs0ny5-_IE9r7Z0r0C:b6r8j5) ([raw file](https://github.com/metaswitch/chronos/blob/0898686bb88400084e7679ca3ea150a63b454012/include/gr_replicator.h#L54)):*
> ```objc
>     _connection(connection),
>     _url(url),
>     _body(body)
> ```

This is a comment I've marked as "Satisfied" - do you still get notified of it sensibly?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/metaswitch/chronos/284)*
<!-- Sent from Reviewable.io -->

rkd-msw commented:
Reviewed 25 of 25 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/timer_handler.cpp, line 712 \[r2\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPs3DRoNcGjSzD387p1:-KPs3DRoNcGjSzD387p2:b-ex1829) ([raw file](https://github.com/metaswitch/chronos/blob/0898686bb88400084e7679ca3ea150a63b454012/src/timer_handler.cpp#L712)):*
> ```cpp
>   std::vector<std::string> old_timer_sites = old_timer->sites;
>   std::vector<std::string> new_timer_sites = new_timer->sites;
>   std::sort(old_timer_sites.begin(), old_timer_sites.end());
> ```

Actually, do you need to sort here? I wonder if you could do something to make the ordering of sites deterministic based on timer ID (e.g. by seeding the random shuffle with the timer ID) - this would then be slightly more efficient.

Not that worried, though.

---

*[src/ut/test_globals.cpp, line 221 \[r2\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPs3wI2E7aoitT0k-uF:-KPs3wI3h6iO_TVl-bTj:b2tqa12) ([raw file](https://github.com/metaswitch/chronos/blob/0898686bb88400084e7679ca3ea150a63b454012/src/ut/test_globals.cpp#L221)):*
> ```cpp
>   expected_remote_site_dns_records.push_back("foo.com:800");
>   expected_remote_site_dns_records.push_back("bar.com");
>   EXPECT_THAT(remote_site_dns_records, UnorderedElementsAreArray(expected_remote_site_dns_records));
> ```

Your expected value comes second in this EXPECT_THAT but first in the other EXPECT_THAT.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/metaswitch/chronos/284)*
<!-- Sent from Reviewable.io -->

eleanor-merry commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[include/gr_replicator.h, line 54 \[r2\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPs0ny4YdbuvPTrMR5p:-KPs5NTeUUoDdqPcX6ro:b2nhswy) ([raw file](https://github.com/metaswitch/chronos/blob/0898686bb88400084e7679ca3ea150a63b454012/include/gr_replicator.h#L54)):*
<details><summary><i>Previously, rkd-msw (Rob Day) wrote…</i></summary>
> This is a comment I've marked as "Satisfied" - do you still get notified of it sensibly?
</details>
Yes

---

*[src/timer_handler.cpp, line 712 \[r2\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPs3DRoNcGjSzD387p1:-KPs5RJpb3HAi_wwPS9h:bfbrotv) ([raw file](https://github.com/metaswitch/chronos/blob/0898686bb88400084e7679ca3ea150a63b454012/src/timer_handler.cpp#L712)):*
<details><summary><i>Previously, rkd-msw (Rob Day) wrote…</i></summary>
> Actually, do you need to sort here? I wonder if you could do something to make the ordering of sites deterministic based on timer ID (e.g. by seeding the random shuffle with the timer ID) - this would then be slightly more efficient.
> 
> Not that worried, though.
</details>
I don't think you can, as we're not changing the ordering of any existing sites (new sites are always added at the end of existing timers)

---

*[src/ut/test_timer_handler.cpp, line 1119 \[r1\]](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KPmC4-b3l4iyf2Dc-4p:-KPs58UNNHQBB4psEE1l:bki1r70) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/src/ut/test_timer_handler.cpp#L1119)):*
<details><summary><i>Previously, rkd-msw (Rob Day) wrote…</i></summary>
> I couldn't immediately see where you'd fixed it, but it looks like the globals were set up in Base::SetUp anyway, right?
</details>
Yup - all the resetting code was unnecessary (and I'm removed it in all the UTs)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/metaswitch/chronos/284)*
<!-- Sent from Reviewable.io -->

rkd-msw commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/metaswitch/chronos/284#-:-KPsBFBwVQ7KGV5-oJjJ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rkd-msw commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 31 files at r1, 1 of 25 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[doc/gr.md, line 15 at r1](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KSH7w9QLm4QQZUMoHwR:-KSH7w9QLm4QQZUMoHwS:bs07t63) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/doc/gr.md#L15)):*
> ```Markdown
>  * Replica 1 - 4 sec delay
>  * Replica 2 - 6 sec delay
> * Site 2
> ```

Hello world

---

*[include/chronos_gr_connection.h, line 28 at r3](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KSH5ajNH82pJeY38vDs:-KSH5ajOeppJLIuIg2AU:b-88houz) ([raw file](https://github.com/metaswitch/chronos/blob/e8ca590cc6c4f64087588348958b0bfbe070bdc0/include/chronos_gr_connection.h#L28)):*
> ```objc
>  * copying, modification, propagation, or distribution would otherwise
>  * violate the terms of the GPL. You must comply with the GPL in all
>  * respects for all of the code used other than OpenSSL.
> ```

FYI, 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/metaswitch/chronos/284#-:-KSH67OaVzvydMIj92tw:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rkd-msw commented:
*[doc/gr.md, line 15 at r1](https://reviewable.io:443/reviews/metaswitch/chronos/284#-KSH7w9QLm4QQZUMoHwR:-KSH87D1TDF-iXn_n80z:b-875oab) ([raw file](https://github.com/metaswitch/chronos/blob/aa8462e8bb625d9a81488dd6a7218d5eddcf170d/doc/gr.md#L15)):*
<details><summary><i>Previously, rkd-msw (Rob Day) wrote…</i></summary>
> Hello world
</details>
Hello

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/metaswitch/chronos/284)*
<!-- Sent from Reviewable.io -->
